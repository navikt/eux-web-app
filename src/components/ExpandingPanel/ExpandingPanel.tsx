import classNames from 'classnames'
import NavHighContrast, { animationClose, animationOpen } from 'nav-hoykontrast'
import { guid } from 'nav-frontend-js-utils'

import PT from 'prop-types'
import React, { useEffect, useState } from 'react'
import { Collapse, UnmountClosed } from 'react-collapse'
import { useTranslation } from 'react-i18next'
import styled from 'styled-components'

export const ExpandingPanelDiv = styled.div`
  &.ekspanderbartPanel--apen .ReactCollapse--collapse {
    will-change: max-height, height;
    max-height: 150em;
    animation: ${animationOpen(150)} 250ms ease;
  }
  &.ekspanderbartPanel--lukket .ReactCollapse--collapse {
    will-change: max-height, height;
    max-height: 0;
    animation: ${animationClose(150)} 250ms ease;
  }
`

export interface ExpandingPanelProps {
  ariaTittel?: string
  border?: boolean
  children : JSX.Element
  className?: string
  collapseProps?: any
  highContrast: boolean
  heading?: JSX.Element | string
  id?: string
  onClose?: () => void
  onOpen?: () => void
  open?: boolean
  renderContentWhenClosed?: boolean
  style?: React.CSSProperties
}

const ExpandingPanel: React.FC<ExpandingPanelProps> = ({
  ariaTittel, border = false, children, className, collapseProps, highContrast, heading, id,
  onClose = () => {}, onOpen = () => {}, open = false, renderContentWhenClosed, style = {}
}: ExpandingPanelProps): JSX.Element => {
  const [_open, setOpen] = useState<boolean>(open)
  const { t } = useTranslation()
  const [_isCloseAnimation, setIsCloseAnimation] = useState<boolean>(false)
  const contentId: string = (collapseProps && collapseProps.id) || guid()

  useEffect(() => {
    if (!open && _open) {
      setIsCloseAnimation(true)
    }
  }, [open, _open])

  const handleOnClick = (e: React.MouseEvent): void => {
    e.preventDefault()
    handleOpenToggle()
  }

  const handleKeyboard = (e: React.KeyboardEvent) => {
    if (e.key === 'Enter') {
      handleOpenToggle()
    }
  }

  const handleOpenToggle = (): void => {
    setOpen(!_open)
    if (!_open) {
      onClose()
    } else {
      onOpen()
    }
  }

  const onRestProxy = (): void => {
    setIsCloseAnimation(false)
    if (collapseProps && collapseProps.onRest) {
      collapseProps.onRest()
    }
  }

  const tabHandler = (e: React.KeyboardEvent) => {
    const { keyCode } = e
    const isTab = keyCode === 9

    if (isTab && _isCloseAnimation) {
      e.preventDefault()
    }
  }

  const showContentId: boolean = !(!renderContentWhenClosed && !_open)
  const ariaControls: {[k: string]: string} | undefined = showContentId ? { 'aria-controls': contentId } : undefined
  const CollapseComponent: any = renderContentWhenClosed ? Collapse : UnmountClosed

  return (
    <NavHighContrast highContrast={highContrast}>
      <ExpandingPanelDiv
        id={id}
        style={style}
        className={classNames('ekspanderbartPanel', className, {
          'ekspanderbartPanel--lukket': !_open,
          'ekspanderbartPanel--apen': _open,
          'ekspanderbartPanel--border': border
        })}
      >
        <div
          aria-expanded={_open}
          data-test-id='c-expandingpanel__head-id'
          className='ekspanderbartPanel__hode'
          onClick={handleOnClick}
          onKeyDown={handleKeyboard}
          role='button'
          tabIndex={0}
          {...ariaControls}
        >
          <div
            className='ekspanderbartPanel__flex-wrapper'
            data-test-id='c-expandingpanel__body-id'
          >
            {heading}
            <button
              aria-expanded={_open}
              aria-label={t('ui:open')}
              className='ekspanderbartPanel__knapp'
              data-test-id='c-expandingpanel__button-id'
              onKeyDown={tabHandler}
              onClick={handleOnClick}
              type='button'
            >
              <span className='ekspanderbartPanel__indikator' />
            </button>
          </div>
        </div>
        <CollapseComponent
          id={contentId}
          isOpened={_open}
          onRest={onRestProxy}
          {...collapseProps}
        >
          <article
            aria-label={ariaTittel}
            className='ekspanderbartPanel__innhold'
            data-test-id='c-expandingpanel__content-id'
          >
            {children}
          </article>
        </CollapseComponent>
      </ExpandingPanelDiv>
    </NavHighContrast>
  )
}

ExpandingPanel.propTypes = {
  ariaTittel: PT.string,
  border: PT.bool,
  children: PT.any.isRequired,
  className: PT.string,
  collapseProps: PT.object,
  highContrast: PT.bool.isRequired,
  heading: PT.oneOfType([PT.string, PT.element]),
  id: PT.string,
  onClose: PT.func,
  onOpen: PT.func,
  open: PT.bool,
  renderContentWhenClosed: PT.bool,
  style: PT.object
}

export default ExpandingPanel
