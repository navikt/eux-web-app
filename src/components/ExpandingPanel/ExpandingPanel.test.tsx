import { mount, ReactWrapper } from 'enzyme'
import React from 'react'
import ExpandingPanel, { ExpandingPanelProps } from './ExpandingPanel'

describe('components/ExpandingPanel/ExpandingPanel', () => {
  let wrapper: ReactWrapper

  const initialMockProps: ExpandingPanelProps = {
    ariaTittel: 'mockAriaTittel',
    border: true,
    collapseProps: {
      id: 'mockId'
    },
    children: <div />,
    heading: <h2>heading</h2>,
    onClick: jest.fn(),
    open: false,
    renderContentWhenClosed: false
  }

  beforeEach(() => {
    wrapper = mount(<ExpandingPanel {...initialMockProps} />)
  })

  afterEach(() => {
    wrapper.unmount()
  })

  it('Renders', () => {
    expect(wrapper.isEmptyRender()).toBeFalsy()
    expect(wrapper).toMatchSnapshot()
  })

  it('Has proper HTML structure', () => {
    expect(wrapper.find('.c-expandingpanel')).toBeTruthy()
    expect(wrapper.find('.ekspanderbartPanel__hode')).toBeTruthy()
    expect(wrapper.find('.ekspanderbartPanel__flex-wrapper')).toBeTruthy()
    expect(wrapper.find('.ekspanderbartPanel__knapp')).toBeTruthy()
    expect(wrapper.find('article.ekspanderbartPanel__innhold')).toBeTruthy()
  })

  it('Opens', () => {
    expect(wrapper.exists('.ekspanderbartPanel--apen')).toBeFalsy()
    expect(wrapper.exists('.ekspanderbartPanel--lukket')).toBeTruthy()
    wrapper.find('.ekspanderbartPanel__knapp').hostNodes().simulate('click')
    wrapper.update()
    expect(wrapper.exists('.ekspanderbartPanel--lukket')).toBeFalsy()
    expect(wrapper.exists('.ekspanderbartPanel--apen')).toBeTruthy()
  })
})
