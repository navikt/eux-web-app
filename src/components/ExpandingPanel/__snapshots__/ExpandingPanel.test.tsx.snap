// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/ExpandingPanel/ExpandingPanel Render: match snapshot 1`] = `
.c1 {
  border-radius: 4px;
  background-color: #FFFFFF;
  border-width: 1px;
  border-style: solid;
  border-color: #78706a;
}

.c1.ekspanderbartPanel--apen .ReactCollapse--collapse {
  will-change: max-height,height;
  max-height: 150em;
  -webkit-animation: lejruX 250ms ease;
  animation: lejruX 250ms ease;
}

.c1.ekspanderbartPanel--lukket .ReactCollapse--collapse {
  will-change: max-height,height;
  max-height: 0;
  -webkit-animation: fywWsj 250ms ease;
  animation: fywWsj 250ms ease;
}

.c1 .ekspanderbartPanel__knapp {
  background: none;
  border: none;
}

.c1.ekspanderbartPanel {
  background-color: #FFFFFF;
}

.c1 .ekspanderbartPanel__hode:focus,
.c1 .ekspanderbartPanel__hode--focus {
  box-shadow: 0 0 0 3px #19548a !important;
  outline: none;
}

.c1 .ekspanderbartPanel__hode {
  background-color: #FFFFFF;
}

.c1 .ekspanderbartPanel__hode:hover {
  background-color: #cce1f3 !important;
}

.c1 .ekspanderbartPanel__indikator {
  color: #3e3832;
}

.c1 p {
  font-size: inherit;
  line-height: inherit;
}

ui-fss.g8[id="c0"] {
  content: "c1,";
}

<ExpandingPanel
  ariaTittel="mockAriaTittel"
  border={true}
  collapseProps={
    Object {
      "id": "mockId",
    }
  }
  heading={
    <h2>
      heading
    </h2>
  }
  highContrast={false}
  onClose={[MockFunction]}
  onOpen={[MockFunction]}
  open={false}
  renderContentWhenClosed={false}
>
  <J
    highContrast={false}
  >
    <Component
      theme={
        Object {
          "alternative-background-color": "#FFFFFF",
          "alternative-hover-color": "#cce1f3",
          "black": "#000000",
          "border-radius-base": "4px",
          "border-radius-large": "6px",
          "border-radius-small": "2px",
          "container-desktop": "(940px + @grid-gutter-width)",
          "container-large-desktop": "(1140px + @grid-gutter-width)",
          "container-lg": "(1140px + @grid-gutter-width)",
          "container-md": "(940px + @grid-gutter-width)",
          "container-sm": "(720px + @grid-gutter-width)",
          "container-tablet": "(720px + @grid-gutter-width)",
          "fokusFarge": "#19548a",
          "grayBackground": "#e9e7e7",
          "grayIcon": "#b7b1a9",
          "grayInactive": "#78706a",
          "grayModia": "#333333",
          "grid-columns": "12",
          "grid-float-breakpoint": "768px",
          "grid-float-breakpoint-max": "(@grid-float-breakpoint - 1)",
          "grid-gutter-width": "16px",
          "inverted-font-color": "#FFFFFF",
          "line-height-large": "1.3333333",
          "line-height-small": "1.5",
          "main-active-color": "#005b82",
          "main-background-color": "#f1f1f1",
          "main-border-color": "#78706a",
          "main-border-radius": "4px",
          "main-border-width": "1px",
          "main-disabled-color": "#b7b1a9",
          "main-error-color": "#ba3a26",
          "main-focus-color": "#19548a",
          "main-font-color": "#3e3832",
          "main-hover-color": "#0074df",
          "main-interactive-color": "#0067c5",
          "main-success-color": "#06893a",
          "main-warning-color": "#ff9100",
          "navBakgrunn": "#ffffff",
          "navBla": "#0067c5",
          "navBlaDarken20": "#32414f",
          "navBlaDarken40": "#254b6d",
          "navBlaDarken60": "#19548a",
          "navBlaDarken80": "#0c5ea8",
          "navBlaLighten20": "#cce1f3",
          "navBlaLighten40": "#99c2e8",
          "navBlaLighten60": "#66a4dc",
          "navBlaLighten80": "#3385d1",
          "navDypBla": "#005b82",
          "navDypBlaDarken20": "#323f42",
          "navDypBlaDarken40": "#254652",
          "navDypBlaDarken60": "#194d62",
          "navDypBlaDarken80": "#0c5472",
          "navDypBlaLighten20": "#ccdee6",
          "navDypBlaLighten40": "#99bdcd",
          "navDypBlaLighten60": "#669db4",
          "navDypBlaLighten80": "#337c9b",
          "navGra20": "#c6c2bf",
          "navGra40": "#b7b1a9",
          "navGra60": "#78706a",
          "navGra80": "#59514b",
          "navGraBakgrunn": "#f1f1f1",
          "navGronn": "#06893a",
          "navGronnDarken20": "#334834",
          "navGronnDarken40": "#285835",
          "navGronnDarken60": "#1c6937",
          "navGronnDarken80": "#117938",
          "navGronnLighten20": "#cde7d8",
          "navGronnLighten40": "#9bd0b0",
          "navGronnLighten60": "#6ab889",
          "navGronnLighten80": "#38a161",
          "navLilla": "#634689",
          "navLillaDarken20": "#453b43",
          "navLillaDarken40": "#4d3e55",
          "navLillaDarken60": "#544066",
          "navLillaDarken80": "#5c4378",
          "navLillaLighten20": "#e0dae7",
          "navLillaLighten40": "#c1b5d0",
          "navLillaLighten60": "#a190b8",
          "navLillaLighten80": "#826ba1",
          "navLimeGronn": "#a2ad00",
          "navLimeGronnDarken20": "#524f28",
          "navLimeGronnDarken40": "#66671e",
          "navLimeGronnDarken60": "#7a7e14",
          "navLimeGronnDarken80": "#8e960a",
          "navLimeGronnLighten20": "#ecefcc",
          "navLimeGronnLighten40": "#dade99",
          "navLimeGronnLighten60": "#c7ce66",
          "navLimeGronnLighten80": "#b5bd33",
          "navLysBla": "#66cbec",
          "navLysBlaDarken20": "#465557",
          "navLysBlaDarken40": "#4e737c",
          "navLysBlaDarken60": "#5690a2",
          "navLysBlaDarken80": "#5eaec7",
          "navLysBlaLighten20": "#e0f5fb",
          "navLysBlaLighten40": "#c2eaf7",
          "navLysBlaLighten60": "#a3e0f4",
          "navLysBlaLighten80": "#85d5f0",
          "navLysGra": "#e9e7e7",
          "navMorkGra": "#3e3832",
          "navOransje": "#ff9100",
          "navOransjeDarken20": "#654a28",
          "navOransjeDarken40": "#8b5c1e",
          "navOransjeDarken60": "#b26d14",
          "navOransjeDarken80": "#d87f0a",
          "navOransjeLighten20": "#ffe9cc",
          "navOransjeLighten40": "#ffd399",
          "navOransjeLighten60": "#ffbd66",
          "navOransjeLighten80": "#ffa733",
          "navRod": "#c30000",
          "navRodDarken20": "#592d28",
          "navRodDarken40": "#73221e",
          "navRodDarken60": "#8e1614",
          "navRodDarken80": "#a80b0a",
          "navRodLighten20": "#f3cccc",
          "navRodLighten40": "#e79999",
          "navRodLighten60": "#db6666",
          "navRodLighten80": "#cf3333",
          "orangeFocus": "#ffbd66",
          "orangeFocusDarken20": "#65533c",
          "orangeFocusDarken40": "#8b6d47",
          "orangeFocusDarken60": "#b28851",
          "orangeFocusDarken80": "#d8a25c",
          "orangeFocusLighten20": "#fff2e0",
          "orangeFocusLighten40": "#ffe5c2",
          "orangeFocusLighten60": "#ffd7a3",
          "orangeFocusLighten80": "#ffca85",
          "padding-base-horizontal": "12px",
          "padding-base-vertical": "6px",
          "padding-large-horizontal": "16px",
          "padding-large-vertical": "10px",
          "padding-small-horizontal": "10px",
          "padding-small-vertical": "5px",
          "padding-xs-horizontal": "5px",
          "padding-xs-vertical": "1px",
          "pinkErrorBg": "#f3e3e3",
          "redError": "#ba3a26",
          "redErrorDarken20": "#573830",
          "redErrorDarken40": "#70392d",
          "redErrorDarken60": "#88392b",
          "redErrorDarken80": "#a13a28",
          "redErrorLighten20": "#f1d8d4",
          "redErrorLighten40": "#e3b0a8",
          "redErrorLighten60": "#d6897d",
          "redErrorLighten80": "#c86151",
          "screen-desktop": "992px",
          "screen-lg": "1200px",
          "screen-lg-desktop": "1200px",
          "screen-lg-min": "1200px",
          "screen-md": "992px",
          "screen-md-max": "(@screen-lg-min - 1)",
          "screen-md-min": "992px",
          "screen-sm": "768px",
          "screen-sm-max": "(@screen-md-min - 1)",
          "screen-sm-min": "768px",
          "screen-tablet": "768px",
          "screen-xs-max": "(@screen-sm-min - 1)",
          "type": "theme",
          "white": "#FFFFFF",
        }
      }
    >
      <styled.div
        className="ekspanderbartPanel ekspanderbartPanel--lukket ekspanderbartPanel--border"
        style={Object {}}
      >
        <div
          className="c0 c1 ekspanderbartPanel ekspanderbartPanel--lukket ekspanderbartPanel--border"
          style={Object {}}
        >
          <div
            aria-expanded={false}
            className="ekspanderbartPanel__hode"
            data-test-id="c-expandingpanel__head-id"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={0}
          >
            <div
              className="ekspanderbartPanel__flex-wrapper"
              data-test-id="c-expandingpanel__body-id"
            >
              <h2>
                heading
              </h2>
              <button
                aria-expanded={false}
                aria-label="ui:open"
                className="ekspanderbartPanel__knapp"
                data-test-id="c-expandingpanel__button-id"
                onClick={[Function]}
                onKeyDown={[Function]}
                type="button"
              >
                <span
                  className="ekspanderbartPanel__indikator"
                />
              </button>
            </div>
          </div>
          <UnmountClosed
            id="mockId"
            isOpened={false}
            onRest={[Function]}
          />
        </div>
      </styled.div>
    </Component>
  </J>
</ExpandingPanel>
`;
